### ISSUE TITLE
Binding numpad keys doesn't work anymore

### ISSUE THREAD

**Issue by stfnm** [i3/i3#2558]
Output of `i3 --moreversion 2>&- || i3 --version`:
```
Binary i3 version:  4.13 (2016-11-08) © 2009 Michael Stapelberg and contributors
Running i3 version: 4.13 (2016-11-08) (pid 566)to abort…)
Loaded i3 config: /home/stfn/.i3/config (Last modified: Sat 12 Nov 2016 02:33:58 PM CET, 721 seconds ago)

The i3 binary you just called: /usr/bin/i3
The i3 binary you are running: i3
```

In my i3 config I have the following configuration:
```
# switch to workspace (with numpad)
bindsym $mod+KP_1 workspace 1
bindsym $mod+KP_2 workspace 2
bindsym $mod+KP_3 workspace 3
bindsym $mod+KP_4 workspace 4
bindsym $mod+KP_5 workspace 5
bindsym $mod+KP_6 workspace 6
bindsym $mod+KP_7 workspace 7
bindsym $mod+KP_8 workspace 8
bindsym $mod+KP_9 workspace 9
bindsym $mod+KP_0 workspace 10
```

The idea is pretty obvious I guess; I simply want to be able to switch i3 workspaces using my modifier key (Windows key) + number keys on my numpad (with numlock enabled).

This used to work fine with i3 versions <= 4.12 but since I have upgraded to 4.13 today it stopped doing anything. When I press for example $mod+0 the 0 is being sent as input to the focused window just like it would if there wasn't any keybinding at all.

My operating system is a fully up-to-date Arch Linux x86-64.

**Comment by i3bot** [comment]
I don’t see a link to logs.i3wm.org. Did you follow http://i3wm.org/docs/debugging.html? (In case you actually provided a link to a logfile, please ignore me.)

**Comment by stapelberg** [comment]
Which keyboard/computer? Which keyboard layout? Can you provide a log file?

I’m somewhat surprised, because we have a testcase which verifies that this feature works (testcases/t/264-keypress-numlock.t).

**Comment by stfnm** [comment]
Here's the complete log as requested:
https://gist.github.com/stfnm/233c3638914f1962399c936408946e31

I pressed $mod+0 at exactly 03:31:30 and waited a bit before and after, doing nothing, so I believe these should be the relevant log entries:

```
11/12/2016 03:31:02 PM - handlers.c:handle_event:1369 - event type 28, xkb_base 85
11/12/2016 03:31:30 PM - handlers.c:handle_event:1369 - event type 85, xkb_base 85
11/12/2016 03:31:30 PM - handlers.c:handle_event:1378 - xkb event, need to handle it.
11/12/2016 03:31:30 PM - handlers.c:handle_event:1404 - xkb state group = 0
11/12/2016 03:31:30 PM - handlers.c:handle_event:1369 - event type 28, xkb_base 85
11/12/2016 03:31:30 PM - handlers.c:handle_event:1369 - event type 28, xkb_base 85
11/12/2016 03:31:30 PM - handlers.c:handle_event:1369 - event type 85, xkb_base 85
11/12/2016 03:31:30 PM - handlers.c:handle_event:1378 - xkb event, need to handle it.
11/12/2016 03:31:30 PM - handlers.c:handle_event:1404 - xkb state group = 0
11/12/2016 03:32:00 PM - handlers.c:handle_event:1369 - event type 7, xkb_base 85
```

My keyboard is a "Das Keyboard Model S : Ultimate EU", my keyboard layout is German and I'm setting it with `setxkbmap -layout de`.

**Comment by klemens** [comment]
I am having the same problem, also under arch with a German keyboard. However this doesn't seem to be relevant, as it also fails with ’setxkbmap -layout us’.

Additionally binding directly to a keycode regressed: I use ’bindcode --release 133 exec rofi’ to start rofi, but also simpler ones like ’bindcode 51 workspace 5’ no longer work. I can open a separate issue if you want, however I thought this may be related.

**Comment by stapelberg** [comment]
Please open a separate issue for that.

**Comment by stef204** [comment]
I can confirm that `bindsym $mod+KP_1 workspace $w1` no longer works, no matter the keyboard layout. 

I have checked the code key again with xev and the above `KP_1` is still correct, etc.

**Comment by bb-q** [comment]
I confirm the issue in 4.13, while 4.12 version has the bindings working. Arch linux, USB keyboard.

**Comment by Dieggho** [comment]
The same with me, I just updated my Arch and all my binding numpad keys stoped working.

**Comment by stapelberg** [comment]
Try adding `Mod2` to the binding, i.e. modify the binding to read `bindsym $mod+Mod2+KP_1 workspace …`.

With changes between v4.12 to v4.13, the numlock fallback became a bit stricter: when the key in question resolves to a different symbol with/without numlock (e.g. KP_1 vs. KP_End), the user needs to explicitly specify the numlock modifier now. This is necessary to disambiguate KP_1 and KP_End — previously, it wasn’t possible to bind on KP_End, but not KP_1.

**Comment by stfnm** [comment]
You're right, that works!

But what do I do with the following bindings?
```
# move focused container to workspace (with numpad)
bindsym $mod+Shift+KP_1 move container to workspace 1
bindsym $mod+Shift+KP_2 move container to workspace 2
bindsym $mod+Shift+KP_3 move container to workspace 3
bindsym $mod+Shift+KP_4 move container to workspace 4
bindsym $mod+Shift+KP_5 move container to workspace 5
bindsym $mod+Shift+KP_6 move container to workspace 6
bindsym $mod+Shift+KP_7 move container to workspace 7
bindsym $mod+Shift+KP_8 move container to workspace 8
bindsym $mod+Shift+KP_9 move container to workspace 9
bindsym $mod+Shift+KP_0 move container to workspace 10
```
Unforunately using `bindsym $mod+Shift+Mod2+KP_1 move container to workspace 1` doesn't seem to do the trick there.

**Comment by stapelberg** [comment]
Does `Shift+KP_1` result in `KP_End` maybe? I don’t have a keyboard with numlock handy to test at the moment.

**Comment by Dieggho** [comment]
Unfortunately Shift+KP_1 doesn't work, control+KP_1 doesn't work as well.

**Comment by stfnm** [comment]
xev tells me Shift+KP_1 does indeed result in KP_End, but in i3 neither `bindsym $mod+Mod2+KP_End move container to workspace 1` nor `bindsym $mod+KP_End move container to workspace 1` appears to do the job.

**Comment by Dieggho** [comment]
I just tested, control+Shift+KP_1 does move the container to workspace 1

**Comment by stfnm** [comment]
I think I've figured it out:
```
# move focused container to workspace (with numpad)
bindsym $mod+Shift+Mod2+KP_End move container to workspace 1
bindsym $mod+Shift+Mod2+KP_Down move container to workspace 2
bindsym $mod+Shift+Mod2+KP_Next move container to workspace 3
bindsym $mod+Shift+Mod2+KP_Left move container to workspace 4
bindsym $mod+Shift+Mod2+KP_Begin move container to workspace 5
bindsym $mod+Shift+Mod2+KP_Right move container to workspace 6
bindsym $mod+Shift+Mod2+KP_Home move container to workspace 7
bindsym $mod+Shift+Mod2+KP_Up move container to workspace 8
bindsym $mod+Shift+Mod2+KP_Prior move container to workspace 9
bindsym $mod+Shift+Mod2+KP_Insert move container to workspace 10
```
Looks rather confusing but they seem to work.

**Comment by bb-q** [comment]
Stapelberg, yes, this works for me. Thanks! What's more strange is that adding Mod2 to the command works with [this issue](https://github.com/i3/i3/issues/2523) too.

**Comment by Crandel** [comment]
My binding "bindsym --release button2 kill" doesn`t work anymore too

**Comment by Airblader** [comment]
@stapelberg So what's the status here… do we close this as works as intended with the grain of salt of having broken setups (something we can't take back anyway)? Or is there something left to do?

**Comment by stef204** [comment]
Having to use `bindsym $mod+Shift+Mod2+KP_Prior move container to workspace 9` etc., looks more like a regression to me, as compared to previous version.  Is there anything which can be done so no breakage?

**Comment by stapelberg** [comment]
For this particular issue (binding keys which are located on your numpad), I don’t think there’s anything we can do.

The change to interpret numlock state literally was a deliberate one, and the unfortunate side-effect of users having to change their keybindings — which is something we try hard to avoid — is the price to pay for having a fully-featured keyboard configuration layer.

I’ll therefore close this issue.

**Comment by stapelberg** [comment]
I’m going to re-open this issue so that people will have an easier time finding it (there have been a couple of duplicates recently).

Regardless of the status, please refer to https://github.com/i3/i3/issues/2558#issuecomment-260770392 — there’s nothing we can do about this issue, and you’ll need to change your keybinding to add `Mod2` in the appropriate place (see the previous comments).

**Comment by tasnad** [comment]
Thanks for the hint with "+Mod2".
However, I still do have an issue with the new scheme.
I have assigned workspaces to the numpad keys and therefore I also need the combination with Shift (for moving windows).
I did some experiments and found that that the binds seem to be inverted with shift:

### Binds without Shift (work as intended)
* bindsym Control+Mod1+**KP_Home** exec beep 1
  -> works with numlock **OFF**
* bindsym Control+Mod1+Mod2+**KP_7** exec beep 2
  -> works with numlock **ON**

### Binds with Shift (KP_Home and KP_7 seem to be inverted)
* bindsym Control+Mod1+Shift+**KP_7** exec beep 3
  -> works with numlock **OFF**
* bindsym Control+Mod1+Shift+Mod2+**KP_Home** exec beep 4
  -> works with numlock **ON**

Should I open a new issue instead?
[i3log.txt](https://github.com/i3/i3/files/606388/i3log.txt)

**Comment by stapelberg** [comment]
Shift does invert the meaning of numpad keys, that’s how the numpad works. Try it out in a text editor, or in `xev(1)`. No need to open a new issue, this is working as intended.

**Comment by stef204** [comment]
@tasnad 
same here re. inversion.  But you can bypass all of that by using `binbcode` instead of `bindym` as I posted above.

**Comment by tasnad** [comment]
@stef204 
Thanks, that does work.
But as I have a python script which generates my config dynamically I'll just add a translation dict for the Keysyms (I like the key Syms) - albeit it's a somewhat hacky solution.

**Comment by rpicot** [comment]
Using bindcode as @stef204 we can use variables to keep something cleanish

```
set $KP_1 87
set $KP_2 88
set $KP_3 89
set $KP_4 83
set $KP_5 84
set $KP_6 85
set $KP_7 79
set $KP_8 80
set $KP_9 81
set $KP_0 90

bindcode $mod+$KP_1 workspace $workspace1
bindcode $mod+$KP_2 workspace $workspace2
bindcode $mod+$KP_3 workspace $workspace3
bindcode $mod+$KP_4 workspace $workspace4
bindcode $mod+$KP_5 workspace $workspace5
bindcode $mod+$KP_6 workspace $workspace6
bindcode $mod+$KP_7 workspace $workspace7
bindcode $mod+$KP_8 workspace $workspace8
bindcode $mod+$KP_9 workspace $workspace9
bindcode $mod+$KP_0 workspace $workspace10

bindcode $mod+Shift+$KP_1 move container to workspace $workspace1
bindcode $mod+Shift+$KP_2 move container to workspace $workspace2
bindcode $mod+Shift+$KP_3 move container to workspace $workspace3
bindcode $mod+Shift+$KP_4 move container to workspace $workspace4
bindcode $mod+Shift+$KP_5 move container to workspace $workspace5
bindcode $mod+Shift+$KP_6 move container to workspace $workspace6
bindcode $mod+Shift+$KP_7 move container to workspace $workspace7
bindcode $mod+Shift+$KP_8 move container to workspace $workspace8
bindcode $mod+Shift+$KP_9 move container to workspace $workspace9
bindcode $mod+Shift+$KP_0 move container to workspace $workspace10
```

**Comment by s3rb31** [comment]
Hello everyone,

I noticed the behaviour and thought it was a bug too. Some thoughts on the topic: 

It might be obvious from a technical point of view to make the user explicitly denote the numlock modifier, since mostly per default it is off.
But from a users perspective it might be confusing because most of the people I know that are using a keyboard with numlock have their settings changed so that numlock is on by default. So I'd argue that KP_1 to KP_9 will be used much more often then KP_Home, KP_End and so on.

So it might be worth flipping it around again. And let people bind with KP_1 to KP_9 and make everyone use the extra Modifier who wants to bind to a numlock-off key.

Greetings

**Comment by Airblader** [comment]
Doing the reverse of what X sees sounds like the perfect idea for confusion to me.

**Comment by tasnad** [comment]
I agree, that would be very confusing. I like the new way of being able to bind numpad keys with respect to the numlock state, so I'd keep it like it is right now.
A solution could be something similar to https://github.com/i3/i3/issues/2577.

**Comment by s3rb31** [comment]
"Doing the reverse of what X sees sounds like the perfect idea for confusion to me."

I agree, maybe I was too quick and didn't thought enough about it. It's just already very confusing plus that Shift inverses the keysym but will still be tested for, makes the confusion even level 2 if you ask me. So I thought getting back to what a user would _expect_ could ease the situation a bit. 

If I overthink, you and tasnad are right because everything happened already. Another change of the the semantics would only do worse. I think I just wanted to point out that not every user has the same insight of the inner structure of what X or i3 see, than a developer.

But then again i3 is targeted at advanced users, so they should know ..

**Comment by Airblader** [comment]
Another crazy idea would be for somebody to write a tool that spits out the correct combination to use. Currently we always refer to `xev` which prints a lot of information the user doesn't need and, in this case, not all the information they need in a meaningful way.

Such a tool could be a great help for a lot of people, I would imagine.

**Comment by tasnad** [comment]
That's not a bad idea!
Meanwhile an overview how to bind commands to the 7 on the numpad (for those who didn't fiddle this out yet):

Numlock ON:
bindsym   Mod2+KP_7                 <cmd1>
bindsym   Shift+Mod2+KP_Home <cmd2>

Numlock OFF:
bindsym   KP_Home   <cmd3>
bindsym   Shift+KP_7 <cmd4>

Four distinct commands are called, depending on the numlock state and whether shift is involved in the keybind.
Hopefully this helps someone.

Cheers,
Tasnad

**Comment by nmschulte** [comment]
> But from a users perspective it might be confusing because most of the people I know that are using a keyboard with numlock have their settings changed so that numlock is on by default. So I'd argue that KP_1 to KP_9 will be used much more often then KP_Home, KP_End and so on.

I just wanted to note about this; I recently received a new laptop to replace my failed one, and it came with the BIOS setting/default for NumLock as on.  It's made by Dell, who in fact seem to have gone out of their way to remove the lock indication LED.  (There is instead, though, an indicator for the microphone mute (cool!); too bad it seems entirely unusable.)
The prior laptop did not have a setting in the supplied BIOS ROM; it is a Clevo machine.

**Comment by jerome-diver** [comment]
I see that this seems to be a kind of chicken fight, but also i hope for someone dev of i3 who be able to recognize objective facts.
For example (it is easy to understand, look...):
KP_END happen why ? because Num_Lock is disable and a key numeric pad at position 1/End is clicked
KP_1 happen why ? because Num_Lock is enable and numeric key pad position 1/End is clicked
where is your problem ? Why add Mod2 ? Why not just recognize it was a terrible bad idea (and not only for logical reason, but also for compatibility). Is there a kind of humility problem to persist a wrong way ?

The SAME physical key is clicked and just because of Num_Lock state On or OFF, the output code change.
So why add this Mod2 (who is, in case of xmodmap is configured has you are thinking) Num_Lock modifier ? There is allready the modification made when press this key...

also, when you read/set the xmodmap, there is 4 choices can be made... that is the original place for do things when shift or not... you just add an overlay where there absolutly no need ot have one more than the existed ones.

You recognize or not a fact doesn't change a fact, but he can increase the quality of your code.

**Comment by Airblader** [comment]
> be able to recognize objective facts

The objective fact is that we made the change deliberately and intentionally and for good reasons that have all been explained in this thread already. To quote yourself,

> You recognize or not a fact doesn't change a fact

I'm also going to close this issue again now (CC @stapelberg), I think it's been long enough and we don't really get issues regarding this anymore.

**Comment by jerome-diver** [comment]
"decision facts" and "objective facts" are two different things.
I well read threads here, and i think you are not read mine.
I definitely not understand why make something more complicate when it serve nothing.
But i can see that is a decision fact you absolutely doesn't want to change, never-mind what users tell you.
enjoy...

**Comment by Airblader** [comment]
> never-mind what users tell you

Without the change, other usecases could not be served and users were complaining that they couldn't use i3 the way they wanted to. With this change, everyone can use it however they want, just the syntax changed in a non-compatible way.

I'm sorry you're not happy with the change, but at least you're not blocked by it in any way, while others previously were.

**Comment by jerome-diver** [comment]
@Airblader , it's ok, i'm happy.
I just feel like it was not necessary (because i not understand), but if you said that it should and that users can access something they can not before... ok then, i just not understand what is it after read all threads.
I'm sorry to insist and maybe the tonality i had wasn't good, i'm not english, so...
don't be hurt, my intentions are not wrong and i'm very happy with i3wm and thank you for this window manager guys.

**Comment by jcelerier** [comment]
If anyone comes here, the configuration that worked for me to switch on $mod + numeric keypad and move $mod + shift + numeric keypad on a french layout is : 

```
bindsym $mod+Mod2+KP_1 workspace 1
bindsym $mod+Mod2+KP_2 workspace 2
bindsym $mod+Mod2+KP_3 workspace 3
bindsym $mod+Mod2+KP_4 workspace 4
bindsym $mod+Mod2+KP_5 workspace 5
bindsym $mod+Mod2+KP_6 workspace 6
bindsym $mod+Mod2+KP_7 workspace 7
bindsym $mod+Mod2+KP_8 workspace 8
bindsym $mod+Mod2+KP_9 workspace 9
bindsym $mod+Mod2+KP_0 workspace 10

bindsym $mod+Shift+Mod2+KP_End move container to workspace 1
bindsym $mod+Shift+Mod2+KP_Down move container to workspace 2
bindsym $mod+Shift+Mod2+KP_Next move container to workspace 3
bindsym $mod+Shift+Mod2+KP_Left move container to workspace 4
bindsym $mod+Shift+Mod2+KP_Begin move container to workspace 5
bindsym $mod+Shift+Mod2+KP_Right move container to workspace 6
bindsym $mod+Shift+Mod2+KP_Home move container to workspace 7
bindsym $mod+Shift+Mod2+KP_Up move container to workspace 8
bindsym $mod+Shift+Mod2+KP_Prior move container to workspace 9
bindsym $mod+Shift+Mod2+KP_Insert move container to workspace 10
```

that's fairly unintuitive but oh well

**Comment by simson** [comment]
@jcelerier Thank you so much !

**Comment by andrei-korshikov** [comment]
Maybe something has been changed, but `Mod2` in `$mod+Shift+Mod2+KP_End` is not needed anymore. Here is my config, works for both (on and off) NumLock statuses.

```
# switch to workspace with digits
bindsym $mod+0          workspace number $ws0
bindsym $mod+1          workspace number $ws1
bindsym $mod+2          workspace number $ws2
bindsym $mod+3          workspace number $ws3
bindsym $mod+4          workspace number $ws4
bindsym $mod+5          workspace number $ws5
bindsym $mod+6          workspace number $ws6
bindsym $mod+7          workspace number $ws7
bindsym $mod+8          workspace number $ws8
bindsym $mod+9          workspace number $ws9

# switch to workspace with keypad when NumLock is off
bindsym $mod+KP_Insert  workspace number $ws0
bindsym $mod+KP_End     workspace number $ws1
bindsym $mod+KP_Down    workspace number $ws2
bindsym $mod+KP_Next    workspace number $ws3
bindsym $mod+KP_Left    workspace number $ws4
bindsym $mod+KP_Begin   workspace number $ws5
bindsym $mod+KP_Right   workspace number $ws6
bindsym $mod+KP_Home    workspace number $ws7
bindsym $mod+KP_Up      workspace number $ws8
bindsym $mod+KP_Prior   workspace number $ws9

# switch to workspace with keypad when NumLock is on (Mod2)
bindsym $mod+Mod2+KP_0  workspace number $ws0
bindsym $mod+Mod2+KP_1  workspace number $ws1
bindsym $mod+Mod2+KP_2  workspace number $ws2
bindsym $mod+Mod2+KP_3  workspace number $ws3
bindsym $mod+Mod2+KP_4  workspace number $ws4
bindsym $mod+Mod2+KP_5  workspace number $ws5
bindsym $mod+Mod2+KP_6  workspace number $ws6
bindsym $mod+Mod2+KP_7  workspace number $ws7
bindsym $mod+Mod2+KP_8  workspace number $ws8
bindsym $mod+Mod2+KP_9  workspace number $ws9

# move focused container to workspace with digits
bindsym $mod+Shift+0            move container to workspace number $ws0
bindsym $mod+Shift+1            move container to workspace number $ws1
bindsym $mod+Shift+2            move container to workspace number $ws2
bindsym $mod+Shift+3            move container to workspace number $ws3
bindsym $mod+Shift+4            move container to workspace number $ws4
bindsym $mod+Shift+5            move container to workspace number $ws5
bindsym $mod+Shift+6            move container to workspace number $ws6
bindsym $mod+Shift+7            move container to workspace number $ws7
bindsym $mod+Shift+8            move container to workspace number $ws8
bindsym $mod+Shift+9            move container to workspace number $ws9

# move focused container to workspace with keypad (NumLock-independent) and switch to it
bindsym $mod+Shift+KP_Insert    move container to workspace number $ws0;    workspace number $ws0
bindsym $mod+Shift+KP_End       move container to workspace number $ws1;    workspace number $ws1
bindsym $mod+Shift+KP_Down      move container to workspace number $ws2;    workspace number $ws2
bindsym $mod+Shift+KP_Next      move container to workspace number $ws3;    workspace number $ws3
bindsym $mod+Shift+KP_Left      move container to workspace number $ws4;    workspace number $ws4
bindsym $mod+Shift+KP_Begin     move container to workspace number $ws5;    workspace number $ws5
bindsym $mod+Shift+KP_Right     move container to workspace number $ws6;    workspace number $ws6
bindsym $mod+Shift+KP_Home      move container to workspace number $ws7;    workspace number $ws7
bindsym $mod+Shift+KP_Up        move container to workspace number $ws8;    workspace number $ws8
bindsym $mod+Shift+KP_Prior     move container to workspace number $ws9;    workspace number $ws9
```

So, `Mod2` is only needed for `KP_0`...`KP_9`. I see the logic, `Mod2` means "NumLock is on", and `KP_0`...`KP_9` exist only when NumLock is on.
